File: lib/utils/prophecy_normalizer.dart
import '../config/prophecy_style.dart';
import '../models/sensor_data.dart';

/// 统一清洗 AI 输出并做质量门判断
class ProphecyNormalizer {
  ProphecyNormalizer._();

  static final _chatMlTokens = RegExp(
    r'<\|im_start\|>(?:system|user|assistant)?|<\|im_end\|>|<\|endoftext\|>',
  );

  static final _assistantPrefix = RegExp(r'^assistant\s*[\n:：]*');

  static final _forbiddenPrefixes = RegExp(
    r'^(预言[：:]\s*|好的[，,]\s*|作为预言家[，,]\s*|我是[^，,。]*[，,]\s*|写一条[^，。]*[，,]\s*)',
  );

  static final _hasChinese = RegExp(r'[\u4e00-\u9fff]');

  static final _hasEnglish = RegExp(r'[a-zA-Z]');

  static final _hasEmoji = RegExp(
    r'[\u{1F300}-\u{1F9FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}]',
    unicode: true,
  );

  /// 模型复述 prompt / 元指令
  static final _promptEcho = RegExp(
    r'上一句[是为：:]|上一句预言|当前传感器|当前状态|传感器数据|写一条|写一句|必须写|编号\d|'
    r'^示例[：:]|^参考[：:]|^数据[：:]|预言家|根据.*数据写|只输出|禁止|要求[：:]',
  );

  static final _promptLeakMarkers = [
    '<|im_start|>',
    '<|im_end|>',
    '传感器数据',
    '当前传感器',
    '当前状态',
    '参考格式',
    '请根据',
    '写一条',
    '写一句',
    '示例',
    '参考：',
    '数据：',
    '上一句',
    '编号',
    '预言：',
    'user\n',
    'assistant\n',
    'system\n',
  ];

  /// 是否像在复述题面或元指令（应拒绝或重试）
  static bool isPromptEcho(String text) {
    final t = text.trim();
    if (t.isEmpty) return true;
    if (_promptEcho.hasMatch(t)) return true;
    if (t.startsWith('数据：') || t.startsWith('参考：')) return true;
    return false;
  }

  /// trim、去 token/前缀/引号，按标点优先截断到 maxChars
  static String normalizeProphecy(String raw) {
    var text = raw.trim();
    if (text.isEmpty) return '';

    text = text.replaceAll(_chatMlTokens, '').trim();
    text = text.replaceAll(_assistantPrefix, '').trim();
    text = text.replaceAll(RegExp(r'[\r\n]+'), '，');
    text = text.replaceAll(RegExp(r'\s{2,}'), '');

    for (final marker in _promptLeakMarkers) {
      if (marker.isEmpty) continue;
      final idx = text.indexOf(marker);
      if (idx > 0) {
        text = text.substring(0, idx).trim();
      }
    }

    text = text.replaceAll(RegExp(r'^["「『]|["」』]$'), '');
    while (_forbiddenPrefixes.hasMatch(text)) {
      text = text.replaceFirst(_forbiddenPrefixes, '').trim();
    }

    text = _stripLeadingMeta(text);
    text = _firstSentence(text);
    text = text.replaceAll(RegExp(r'[，、；]+$'), '');
    if (text.isEmpty) return '';

    if (text.length <= ProphecyStyle.maxChars) return text;
    return _truncateAtPunctuation(text, ProphecyStyle.maxChars);
  }

  static String _stripLeadingMeta(String text) {
    final patterns = [
      RegExp(r'^上一句[是为：:]\s*'),
      RegExp(r'^当前状态[是为：:]\s*'),
      RegExp(r'^当前传感器[：:]\s*'),
      RegExp(r'^数据[：:]\s*'),
      RegExp(r'^参考[：:]\s*'),
    ];
    for (final p in patterns) {
      if (p.hasMatch(text)) {
        text = text.replaceFirst(p, '').trim();
      }
    }
    return text;
  }

  static String _firstSentence(String text) {
    final end = RegExp(r'[。！？]');
    final match = end.firstMatch(text);
    if (match != null) {
      final candidate = text.substring(0, match.end).trim();
      if (candidate.length >= ProphecyStyle.minChars) return candidate;
    }
    return text.trim();
  }

  static String _truncateAtPunctuation(String text, int maxLen) {
    final punct = RegExp(r'[，。！？、；]');
    var best = maxLen;
    for (var i = 0; i < text.length && i < maxLen; i++) {
      if (punct.hasMatch(text[i])) best = i + 1;
    }
    if (best < ProphecyStyle.minChars) best = maxLen;
    return text.substring(0, best.clamp(0, text.length));
  }

  /// 长度、中文、无违规前缀（不必含传感器数据）
  static bool isAcceptableProphecy(String text) =>
      isSoftAcceptableProphecy(text);

  /// MLX 专用：清洗后符合基本废话格式即可
  static bool isMlxProphecy(String text, SensorData sensor) {
    final cleaned = normalizeProphecy(text);
    return isSoftAcceptableProphecy(cleaned);
  }

  /// 放宽门：有中文、长度合规即可
  static bool isSoftAcceptableProphecy(String text) {
    if (text.isEmpty || isPromptEcho(text)) return false;
    if (text.length < ProphecyStyle.minChars ||
        text.length > ProphecyStyle.maxChars) {
      return false;
    }
    if (!_hasChinese.hasMatch(text)) return false;
    if (_hasEnglish.hasMatch(text)) return false;
    if (_hasEmoji.hasMatch(text)) return false;
    if (_forbiddenPrefixes.hasMatch(text)) return false;
    return true;
  }
}

File: server/app/services/prophecy_normalizer.py
"""预言输出清洗与质量门（移植自 lib/utils/prophecy_normalizer.dart）。"""

from __future__ import annotations

import re

MAX_CHARS = 45
MIN_CHARS = 12

_CHAT_ML_TOKENS = re.compile(
    r"<\|im_start\|>(?:system|user|assistant)?|<\|im_end\|>|<\|endoftext\|>"
)
_ASSISTANT_PREFIX = re.compile(r"^assistant\s*[\n:：]*")
_FORBIDDEN_PREFIXES = re.compile(
    r"^(预言[：:]\s*|好的[，,]\s*|作为预言家[，,]\s*|我是[^，,。]*[，,]\s*|写一条[^，。]*[，,]\s*)"
)
_HAS_CHINESE = re.compile(r"[\u4e00-\u9fff]")
_HAS_ENGLISH = re.compile(r"[a-zA-Z]")
_HAS_EMOJI = re.compile(
    r"[\U0001F300-\U0001F9FF\u2600-\u26FF\u2700-\u27BF]"
)
_PROMPT_ECHO = re.compile(
    r"上一句[是为：:]|上一句预言|当前传感器|当前状态|传感器数据|写一条|写一句|必须写|编号\d|"
    r"^示例[：:]|^参考[：:]|^数据[：:]|预言家|根据.*数据写|只输出|禁止|要求[：:]"
)
_PROMPT_LEAK_MARKERS = [
    "<|im_start|>",
    "<|im_end|>",
    "传感器数据",
    "当前传感器",
    "当前状态",
    "参考格式",
    "请根据",
    "写一条",
    "写一句",
    "示例",
    "参考：",
    "数据：",
    "上一句",
    "编号",
    "预言：",
    "user\n",
    "assistant\n",
    "system\n",
]
_STRIP_META_PATTERNS = [
    re.compile(r"^上一句[是为：:]\s*"),
    re.compile(r"^当前状态[是为：:]\s*"),
    re.compile(r"^当前传感器[：:]\s*"),
    re.compile(r"^数据[：:]\s*"),
    re.compile(r"^参考[：:]\s*"),
]
_PUNCT_END = re.compile(r"[。！？]")
_PUNCT_TRUNC = re.compile(r"[，。！？、；]")


def is_prompt_echo(text: str) -> bool:
    t = text.strip()
    if not t:
        return True
    if _PROMPT_ECHO.search(t):
        return True
    return t.startswith("数据：") or t.startswith("参考：")


def normalize_prophecy(raw: str) -> str:
    text = raw.strip()
    if not text:
        return ""

    text = _CHAT_ML_TOKENS.sub("", text).strip()
    text = _ASSISTANT_PREFIX.sub("", text).strip()
    text = re.sub(r"[\r\n]+", "，", text)
    text = re.sub(r"\s{2,}", "", text)

    for marker in _PROMPT_LEAK_MARKERS:
        if not marker:
            continue
        idx = text.find(marker)
        if idx > 0:
            text = text[:idx].strip()

    text = re.sub(r'^["「『]|["」』]$', "", text)
    while _FORBIDDEN_PREFIXES.match(text):
        text = _FORBIDDEN_PREFIXES.sub("", text, count=1).strip()

    for pattern in _STRIP_META_PATTERNS:
        if pattern.match(text):
            text = pattern.sub("", text, count=1).strip()

    text = _first_sentence(text)
    text = re.sub(r"[，、；]+$", "", text)
    if not text:
        return ""

    if len(text) <= MAX_CHARS:
        return text
    return _truncate_at_punctuation(text, MAX_CHARS)


def _first_sentence(text: str) -> str:
    match = _PUNCT_END.search(text)
    if match:
        candidate = text[: match.end()].strip()
        if len(candidate) >= MIN_CHARS:
            return candidate
    return text.strip()


def _truncate_at_punctuation(text: str, max_len: int) -> str:
    best = max_len
    for i, ch in enumerate(text[:max_len]):
        if _PUNCT_TRUNC.match(ch):
            best = i + 1
    if best < MIN_CHARS:
        best = max_len
    return text[: min(best, len(text))]


def is_soft_acceptable_prophecy(text: str) -> bool:
    if not text or is_prompt_echo(text):
        return False
    if len(text) < MIN_CHARS or len(text) > MAX_CHARS:
        return False
    if not _HAS_CHINESE.search(text):
        return False
    if _HAS_ENGLISH.search(text):
        return False
    if _HAS_EMOJI.search(text):
        return False
    if _FORBIDDEN_PREFIXES.match(text):
        return False
    return True
